File: core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/pythonworker/promisehandlers/EvaluateExpressionHandler.scala
package edu.uci.ics.amber.engine.architecture.pythonworker.promisehandlers

import edu.uci.ics.amber.engine.architecture.worker.controlreturns.EvaluatedValue
import edu.uci.ics.amber.engine.common.rpc.AsyncRPCServer.ControlCommand

object EvaluateExpressionHandler {
  final case class EvaluateExpression(expression: String) extends ControlCommand[EvaluatedValue]
}

File: core/amber/src/main/python/core/architecture/handlers/evaluate_expression_handler.py
from proto.edu.uci.ics.amber.engine.architecture.worker import (
    EvaluateExpressionV2,
    EvaluatedValue,
)
from .handler_base import Handler
from ..managers.context import Context
from ...util.expression_evaluator import ExpressionEvaluator


class EvaluateExpressionHandler(Handler):
    cmd = EvaluateExpressionV2

    def __call__(self, context: Context, command: cmd, *args, **kwargs):
        runtime_context = {
            r"self": context.operator_manager.operator,
            r"tuple_": context.tuple_processing_manager.current_input_tuple,
            r"input_": context.tuple_processing_manager.current_input_link,
        }

        evaluated_value: EvaluatedValue = ExpressionEvaluator.evaluate(
            command.expression, runtime_context
        )

        return evaluated_value
